### PR TITLE
Fix and tests added DAFFODIL-2873 zoned ebcdic bug

### PR DESCRIPTION
**Pull request by mbeckerle**
DAFFODIL-2873

### CODE REVIEW

File: daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        <maxInclusive value="9999999.99"/>
+        <minInclusive value="-9999999.99"/>
+      </restriction>
+    </simpleType>
+
+    <element name="money7" type="ex:displayMoney7Digits"/>
+
+  </tdml:defineSchema>
+
+  <tdml:parserTestCase name="ZonedEBCDICLeadingSign" root="money7" model="ebcdic1"
```
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml:72]: tdml: prefixes can all be removed.

File: daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    </simpleType>
+
+    <element name="money7" type="ex:displayMoney7Digits"/>
+
+  </tdml:defineSchema>
+
+  <parserTestCase name="ZonedEBCDICLeadingSign" root="money7" model="ebcdic1"
+    description="Test of a negative number with ebcdic charset, leading overpunched sign of R should be -9">
+    <document>
+      <documentPart type="text" encoding="ebcdic-cp-us">R99999999</documentPart>
```
Comment (stevedlawrence) [inline daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml:75]: It's there any value in defining the document part in bytes instead of text with ebcdic-cp-us encoding? That would show that the issue doesn't have anything to do with ebcdic encoding in the documentPartt? I feel like that isn't used very often so there's a chance the bug is in that?

**Review by stevedlawrence — APPROVED** [review]
+1 needs  `sbt scalafmtAll` to quiet the linter.

File: daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/RuntimePropertyMixins.scala [hunk 3]
```diff
@@ -123,7 +123,7 @@ trait TermRuntimeValuedPropertiesMixin
     ev
   }
 
-  final lazy val charsetEv = {
+  final lazy val charsetEv: CharsetEv = {
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/RuntimePropertyMixins.scala:126]: Revert

File: daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala [hunk 4]
```diff
@@ -700,6 +701,30 @@ trait ElementBaseGrammarMixin
...
+  /**
+   * True if the encoding is known to be an EBCDIC one, as in the encoding is not  a runtime expression
+   * and it's some ebcdic flavor. If it's any ascii flavor or a runtime expression this is false.
+   */
+  lazy val isKnownEBCDICEncoding: Boolean = {
+    charsetEv.optConstant
+      .map { bcs: BitsCharset =>
+        val nom = bcs.name.toUpperCase()
+        val res = (nom == "IBM037") ||
+          nom.startsWith("EBCDIC")
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:713]: This may not be sufficient to recognize EBCDIC. There are numrous IBM code pages that are EBCDIC besides IBM037. There are ones for the international variants of EBCDIC. All of them do the exact same thing for zoned numbers. 

We need something that will work regardless of the aliases vs. the "official" name. If the name, or any alias contains "ebcdic" (case insensitive) then it is an ebcdic variant.

Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:713]: Should we add an `isEBCDIC` member to `BitsCharset` that defaults to `false`, and any EBCDIC encodings we implement would override that to `true`? Then this just becomes

```scala
charsetEv.optConstant.map { _.isEBCDIC }.getOrElse(false)
```

Also, doing some grepping for EBCDIC I see we also have a `Misc.isAsciiBased` function. That's only used for dumping debug data and works on Java charsets instead of our BitsCharset, but maybe it provides extra motivation that we might eventually want to have other `is*` members so we can ask questions about our supported charsets without having to make guesses based on the name.

File: daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala [hunk 5]
```diff
@@ -700,6 +701,30 @@ trait ElementBaseGrammarMixin
...
+  }
+
+  /**
+   * Avoids requesting the textZonedSignStyle property if we know the encoding
+   * is an EBCDIC flavor.
+   */
+  lazy val optTextZonedSignStyle = {
+    if (isKnownEBCDICEncoding) None
+    else Some(textZonedSignStyle)
+  }
```
Comment (stevedlawrence) [inline daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:726]: This doesn't do the right thing if the charset is non-constant be evaluates to EBCIDC at runtime. Is that something we care about? Does it make sense to convert this to a evaluatable?

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala [hunk 6]
```diff
@@ -386,6 +386,25 @@ object DecimalUtils {
...
+   */
+  def convertFromZonedEBCDIC(digit: Char): (Int, Boolean) = {
+    if ((digit >= '0') && (digit <= '9')) // positive 0-9
+      (digit - 48, false)
+    else if (digit == '}') // negative 0
+      (0, true)
+    else if ((digit >= 'J') && (digit <= 'R')) // negative 1-9
+      (digit - 'J' + 1, true)
+    else
+      throw new NumberFormatException("Invalid zoned digit: " + digit)
```
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala:405]: The DFDL spec says this about `textNumberRep`:
> In EBCDIC-based encodings, code points 0xC0 to 0xC9 or 0xF0 to 0xF9 represent a positive sign and digits 0 to 9 (these byte ranges correspond typically to characters '{ABCDEFGHI'* or '0123456789'), and code points 0xD0 to 0xD9 or 0xB0 to 0xB9 represent a negative sign and digits 0 to 9 (these byte ranges correspond typically to characters '}JKLMNOPQR' or  '^£¥·©§¶¼½¾ ' ). On parsing both ranges are accepted. On unparsing the range 0xC0 to 0xC9 are produced for positive signs and the range 0xD0 to 0xD9 are produced for negative signs.

Should this function also support `{ABCDEFGHI` for positive numbers and `^£¥·©§¶¼½¾ ` for negatives?

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala [hunk 7]
```diff
@@ -417,6 +436,25 @@ object DecimalUtils {
       throw new NumberFormatException("Invalid zoned digit: " + digit)
   }
 
+  /**
+   * Does not encode to the EBCDIC charset, but converts negative digits
+   * to their EBCDIC Zoned overpunched representation where negative 0 to 9
+   * are mapped to "}JKLMNOPQR".
+   * @param digit - the digit, as a char '0' to '9'
+   * @param positive - true if positive, false if negative
+   * @return the character needed to represent this character as an overpunched sign digit.
```
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala:445]: From the spec:

> In EBCDIC-based encodings, code points 0xC0 to 0xC9 or 0xF0 to 0xF9 represent a positive sign and digits 0 to 9 (these byte ranges correspond typically to characters '{ABCDEFGHI' or '0123456789'), and code points 0xD0 to 0xD9 or 0xB0 to 0xB9 represent a negative sign and digits 0 to 9 (these byte ranges correspond typically to characters '}JKLMNOPQR' or  '^£¥·©§¶¼½¾ ' ). On parsing both ranges are accepted. On unparsing the range 0xC0 to 0xC9 are produced for positive signs and the range 0xD0 to 0xD9 are produced for negative signs.

This makes it sound like on unparsing the positive 0-9 digits should be unparsed as `{ABCDFGHI` instead of 0-9?

File: daffodil-lib/src/test/scala/org/apache/daffodil/lib/util/TestDecimalUtils.scala [hunk 8]
```diff
@@ -1846,733 +1989,1620 @@ class TestDecimalUtils {
...
-      zonedFromNumber("-9", TextZonedSignStyle.AsciiTandemModified, OverpunchLocation.Start),
+      zonedFromNumber(
+        "-9",
+        Some(TextZonedSignStyle.AsciiTandemModified),
+        OverpunchLocation.Start,
+      ),
       "",
     )
   }
+
```
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-lib/src/test/scala/org/apache/daffodil/lib/util/TestDecimalUtils.scala:3201]: Same question as above, should the positives be `{` and `A-I` instead of `0-9`? And do we need tests for the other range of positive/negative values?

**Review by stevedlawrence — APPROVED** [review]
+1

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala (hunk 7 above)
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala:466]: You are correct. Fixed.

File: daffodil-lib/src/test/scala/org/apache/daffodil/lib/util/TestDecimalUtils.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-lib/src/test/scala/org/apache/daffodil/lib/util/TestDecimalUtils.scala:3201]: You are correct. Fixed and tests added.

File: daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala (hunk 6 above)
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala]: Correct. Fixed.

**Comment by mbeckerle** [comment]
I believe this now addresses all the comments by @stevedlawrence.

Needs a second reviewer. @jadams-tresys can you review this? I believe the ascii zoned stuff was first done by you.

**Comment by jadams-tresys** [comment]
+1

Looks good to me.  Reading the Wikipedia page I'm not surprised I didn't get this right on the first try.  This section was a fun read though: https://en.wikipedia.org/wiki/EBCDIC#Criticism_and_humor

**Review by jadams-tresys — APPROVED** [review]

File: daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala (hunk 5 above)
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:717]: But if the charset evaluates to EBCDIC at runtime we'll still use textZonedSignStyle. So in that case we still need the property at compile time (like you have) but we need to ignore it at runtime.

So maybe the parser/unparsers need to have the charsetEv passed in, and if at runtime if it's non-constant and evaluates to EBCDIC then it ignores `optTextZonedSignStyle`?

Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:717]: Correct. That is what happens. So this optional thing is just about not reqiring what is an ascii-oriented property at all in the case where it is known for certain to be EBCDIC.
